runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/streaming/harness/WindmillStreamSender.java

```diff
@@ -154,18 +156,13 @@ void closeAllStreams() {
   }
 
   @Override
-  public void adjustBudget(long itemsDelta, long bytesDelta) {
-    getWorkBudget.set(getWorkBudget.get().apply(itemsDelta, bytesDelta));
+  public void setBudget(long items, long bytes) {
+    getWorkBudget.set(getWorkBudget.get().apply(items, bytes));
     if (started.get()) {
```

Comment:
Since `started` is set inside startStreams after `getWorkStream.get()` without mutexes guarding them, `started` can be false here setBudget is called by a different thread when `startStreams` is in the middle of execution. do we need to change the synchronization in this class?

Reply:
that is in this PR https://github.com/apache/beam/pull/32774 and i added synchronization there
this is still the memoized Supplier.get() which is threadsafe